smac/intensification/successive_halving.py

```diff
@@ -0,0 +1,462 @@
+import time
+import logging
+import typing
+from collections import OrderedDict
+from itertools import islice
+
+import numpy as np
+
+from smac.intensification.intensification import Intensifier
+from smac.optimizer.objective import sum_cost
+from smac.stats.stats import Stats
+from smac.utils.constants import MAXINT
+from smac.configspace import Configuration
+from smac.runhistory.runhistory import RunHistory
+from smac.tae.execute_ta_run import BudgetExhaustedException, CappedRunException, ExecuteTARun
+from smac.utils.io.traj_logging import TrajLogger
+
+__author__ = "Ashwin Raaghav Narayanan"
+__copyright__ = "Copyright 2018, ML4AAD"
+__license__ = "3-clause BSD"
+
+
+class SuccessiveHalving(Intensifier):
+    """Races multiple challengers against an incumbent using Successive Halving method
+
+    Implementation from "BOHB: Robust and Efficient Hyperparameter Optimization at Scale" (Falkner et al. 2018)
+    (refer supplementary)
+
+    Parameters
+    ----------
+    tae_runner : smac.tae.execute_ta_run.ExecuteTARun Object
+        target algorithm run executor
+    stats: smac.stats.stats.Stats
+        stats object
+    traj_logger: smac.utils.io.traj_logging.TrajLogger
+        TrajLogger object to log all new incumbents
+    rng : np.random.RandomState
+    instances : typing.List[str]
+        list of all instance ids
+    instance_specifics : typing.Mapping[str,np.ndarray]
+        mapping from instance name to instance specific string
+    cutoff : typing.Union[int, None]
+        cutoff of TA runs
+    deterministic : bool
+        whether the TA is deterministic or not
+    min_budget : typing.Union[float, None]
```

Comment:
Could you please rename this to `initial_budget` as it is called in Algorithm 1 of the Appendix of Falkner et al.?

Reply:
In "hpbandster" implementations they had named it as `min_budget`. So I thought it would be easier to relate between the two implementations this way. If that is not needed, I'll change it.